database/pgx/pgx.go

```diff
@@ -298,9 +298,21 @@ func (p *Postgres) runStatement(statement []byte) error {
 			if pgErr.Detail != "" {
 				message = fmt.Sprintf("%s, %s", message, pgErr.Detail)
 			}
-			return database.Error{OrigErr: err, Err: message, Query: statement, Line: line}
+			migrationErr = database.Error{OrigErr: err, Err: message, Query: statement, Line: line}
+		} else {
+			migrationErr = database.Error{OrigErr: err, Err: "migration failed", Query: statement}
+		}
+
+		// For safety, always issue a rollback on error. In multi-statement
+		// mode, this is necessary to make sure that the connection is not left
+		// in an aborted state. In single-statement mode, this will be a no-op
+		// outside of the implicit transaction block that was already rolled
+		// back.
+		if _, rollbackErr := p.conn.ExecContext(ctx, "ROLLBACK"); rollbackErr != nil {
+			rollbackErr = fmt.Errorf("failed to rollback migration tx: %w", rollbackErr)
+			return multierror.Append(migrationErr, rollbackErr)
```

Comment:
Here's a simple example of what this output might look like from a basic console logger in the case where both the migration statement _and_ this explicit rollback generated an error:

```
2023/09/20 15:20:30 Database connection established
2023/09/20 15:20:30 Error running migrations: 2 errors occurred:
        * migration failed: division by zero in line 0:
SELECT 1/0; -- Divide by zero error!

 (details: pq: division by zero)
        * failed to rollback migration tx: pq: column "doesnt-exist" does not exist

2023/09/20 15:20:30 Closing database connection
```

I aimed to make this message clear that what failed wasn't a "rollback of the migrate schema version itself" , but just a rollback of the transaction being used in the up migration that was being attempted. Happy to take feedback on this error to make that clearer, though!

Reply:
My only nit on the wrapped error message is that `failed to` is kinda noise, as suggested by https://github.com/uber-go/guide/blob/master/style.md#error-wrapping, but it looks like this repo uses `Failed to...` already https://github.com/search?q=repo%3Anicheinc%2Fmigrate+Errorf+%25w&type=code

Reply:
Fine by me! In https://github.com/nicheinc/migrate/pull/1/commits/bfdd2f0b5fe1c6952d155c5f49f5d9c2f24a396b I changed the wording from "failed to rollback migration tx: " to "rolling back migration tx: ".